src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2793,13 +2812,47 @@ private void InheritDefaultState(TypeSymbol targetType, int targetSlot)
             {
                 var symbol = AsMemberOfType(targetType, variable.Symbol);
                 SetStateAndTrackForFinally(ref this.State, slot, GetDefaultState(symbol));
-                InheritDefaultState(symbol.GetTypeOrReturnType().Type, slot);
+                InheritDefaultState(GetTypeOrReturnType(symbol), slot);
             }
             members.Free();
         }
 
+        private static TypeSymbol GetTypeOrReturnType(Symbol symbol) => symbol.GetTypeOrReturnType().Type;
+
+        /// <summary>Gets the TypeWithAnnotations of a symbol, possibly using the inferred nullable annotation for backing fields.</summary>
+        private TypeWithAnnotations GetTypeOrReturnTypeWithAnnotations(Symbol symbol)
+        {
+            var typeWithAnnotations = symbol.GetTypeOrReturnType();
+            if (symbol is SynthesizedBackingFieldSymbol { InfersNullableAnnotation: true } backingField)
+            {
+                NullableAnnotation nullableAnnotation;
+                if (_getterNullResilienceData is var (analyzedField, assumedNullableAnnotation)
+                    && (object)analyzedField == backingField)
```

Comment:
>(object)analyzedField == backingField [](http://example.com/codeflow?start=23&length=37)

Does this ever fail? If not, should we assert this somewhere?

Reply:
I wasn't able to think of any existing code where this fails. I believe this could only happen if a getter is using a backing field of another property. Then, when we are inferring annotation of the current backing field, we would start inferring an annotation of a different backing field.

The 'else' block effectively asserts this, though, I am open to some more straightforward way of writing this.

I do think it is good to keep doing this check in release mode, though. I actually wouldn't mind just throwing here in release mode if this is not met, and dropping the assertion.

Reply:
Done